clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -0,0 +1,638 @@
+     * on the preference).
+     * @return Non-empty map of supported protocols and metadata
+     */
+    protected abstract LinkedHashMap<String, ByteBuffer> metadata();
+
+    /**
+     * Invoked when a group member has successfully joined a group.
+     * @param generation The generation that was joined
+     * @param memberId The identifier for the local member in the group
+     * @param protocol The protocol selected by the coordinator
+     * @param memberAssignment The assignment propagated from the group leader
+     */
+    protected abstract void onJoin(int generation,
+                                   String memberId,
+                                   String protocol,
+                                   ByteBuffer memberAssignment);
+
+    /**
+     * Perform synchronization for the group. This is used by the leader to push state to all the members
+     * of the group (e.g. to push partition assignments in the case of the new consumer)
+     * @param leaderId The id of the leader (which is this member)
+     * @param allMemberMetadata Metadata from all members of the group
+     * @return A map from each member to their state assignment
+     */
+    protected abstract Map<String, ByteBuffer> doSync(String leaderId,
+                                                      String protocol,
+                                                      Map<String, ByteBuffer> allMemberMetadata);
+
+    /**
+     * Invoked when the group is left (whether because of shutdown, metadata change, stale generation, etc.)
+     * @param generation The generation that was left
+     * @param memberId The identifier of the local member in the group
+     */
+    protected abstract void onLeave(int generation, String memberId);
+
+
+    /**
+     * Block until the coordinator for this group is known.
+     */
+    public void ensureCoordinatorKnown() {
+        while (coordinatorUnknown()) {
+            RequestFuture<Void> future = sendGroupMetadataRequest();
+            client.poll(future, requestTimeoutMs);
+
+            if (future.failed())
+                client.awaitMetadataUpdate();
+        }
+    }
+
+    /**
+     * Check whether the group should be rejoined (e.g. if metadata changes)
+     * @return true if it should, false otherwise
+     */
+    protected boolean needRejoin() {
+        return rejoinNeeded;
+    }
+
+    /**
+     * Reset the generation/memberId tracked by this member
+     */
+    public void resetGeneration() {
+        this.generation = OffsetCommitRequest.DEFAULT_GENERATION_ID;
+        this.memberId = JoinGroupRequest.UNKNOWN_MEMBER_ID;
+        rejoinNeeded = true;
+    }
+
+    /**
+     * Ensure that the group is active (i.e. joined and synced)
+     */
+    public void ensureActiveGroup() {
+        if (!needRejoin())
+            return;
+
+        // onLeave only invoked if we have a valid current generation
+        onLeave(generation, memberId);
```

Comment:
@hachikuji Good point about the correlation with rebalance listener methods. You are right, we should make them consistent. 

It looks we are trying to address two different kinds of programmers:
1. People who are using KafkaConsumer.
2. People who are developing their clients(producer/consumer/processor, etc) and extending AbstractCoordinator.

Majority of the programmers belongs to (1). What they need to provides today are:
* ConsumerRebalanceListener
  * `onParitionsRevoked()`
  * `onParitionsAssigned()`
* PartitionAssignor
  * `assign()`
  * `onAssignment()`
  * `subscription()`

The programmers in case (2) need to think about:
* `onGroupJoinStart()`
* `generateAssignment()`
* `onGroupJoinEnd()`

To make everything consistent, perhaps we can change the ConsumerRebalanceListener interface to:
* ConsumerRebalanceListener
  * `onGroupJoinStart()`
  * `onGroupJoinEnd()`

As said in KAFKA-2674, my only concern about this approach is that if later on we want to add a `beforeCommittingOffset()` to rebalance listener, that's going to be a bit weird. But there is always a workaround to turn off auto commit and let user have full control over the state when `onGroupJoinStart` is called. So I guess this is less a problem.

Does this approach make sense?

BTW, WRT **leave**, here is my understandings about onLeave() if we have one.
If we have an onLeave(), it should only be called when a consumer has been kicked out of the group, i.e. its generation id is no longer valid. Otherwise, the consumer is technically still in the group. That means onLeave() should not be called during a normal rebalance because the generation id is still valid.